internals/overlord/planstate/manager.go

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2024 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package planstate
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/plan"
+)
+
+// LabelExists is the error returned by AppendLayer when a layer with that
+// label already exists.
+type LabelExists struct {
+	Label string
+}
+
+func (e *LabelExists) Error() string {
+	return fmt.Sprintf("layer %q already exists", e.Label)
+}
+
+type PlanManager struct {
+	state     *state.State
+	runner    *state.TaskRunner
+	pebbleDir string
+
+	planLock     sync.Mutex
+	plan         *plan.Plan
+	planHandlers []PlanChangedFunc
+}
+
+func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string) (*PlanManager, error) {
+	manager := &PlanManager{
+		state:     s,
+		runner:    runner,
+		pebbleDir: pebbleDir,
+		plan:      &plan.Plan{},
+	}
+
+	return manager, nil
+}
+
+// Load reads plan layers from the pebble directory, combines and validates the
+// final plan, and finally notifies registered managers of the plan update. In
+// the case of a non-existent layers directory, or no layers in the layers
+// directory, an empty plan is announced to change subscribers.
+func (m *PlanManager) Load() error {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+	plan, err := plan.ReadDir(m.pebbleDir)
+	if err != nil {
+		return err
+	}
+	m.planChanged(plan)
+	return nil
+}
+
+// PlanChangedFunc is the function type used by AddChangeListener.
+type PlanChangedFunc func(p *plan.Plan)
+
+// AddChangeListener adds f to the list of functions that are called whenever
+// a plan change event took place (Load, AppendLayer, CombineLayer). A plan
+// change event does not guarantee that combined plan content has changed.
+// Notification registration must be completed before the plan is loaded.
+func (m *PlanManager) AddChangeListener(f PlanChangedFunc) {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+	m.planHandlers = append(m.planHandlers, f)
+}
+
+func (m *PlanManager) planChanged(plan *plan.Plan) {
+	m.plan = plan
+	for _, f := range m.planHandlers {
+		f(plan)
+	}
+}
+
+// Plan returns the combined configuration plan. Any change made to the plan
+// will result in a new Plan instance, so the current design assumes a returned
+// plan is never mutated by planstate (and may never be mutated by any
+// consumer).
+func (m *PlanManager) Plan() (*plan.Plan, error) {
```

Comment:
Can we just return the `*Plan` and no error? This can never return an error now, by design, so it seems unnecessary/awkward to have the caller have to check one.
```suggestion
func (m *PlanManager) Plan() *plan.Plan {
```

Reply:
Improved. Thank you. Resolving.